server/app/models/grid_service.rb

```diff
@@ -244,12 +244,18 @@ def dependant_services
     grid = self.grid
     dependant = []
     dependant += grid.grid_services.where(:$or => [
-        {:volumes_from => {:$regex => /^#{self.name}-%s/}},
-        {:volumes_from => {:$regex => /^#{self.name}-\d+/}},
-        {:affinity => "service==#{self.name}"},
-        {:affinity => "service!=#{self.name}"},
-        {:net => {:$regex => /^container:#{self.name}-%s/}},
-        {:net => {:$regex => /^container:#{self.name}-\d+/}}
+        # if this is a stackless service, this also includes bare volumes_from service names in other stacks
+        {:volumes_from => {:$regex => /^#{self.name_with_stack}-(%s|\d+)/}},
+
+        # the agent will resolve volumes_from: foo-1 to container stack.foo-1 if container foo-1 does not exist
+        {:volumes_from => {:$regex => /^#{self.name}-(%s|\d+)/}, :stack_id => self.stack.id},
+
+        {:affinity => "service==#{self.name}", :stack_id => self.stack.id},
+        {:affinity => "service!=#{self.name}", :stack_id => self.stack.id},
+        {:affinity => "service==#{self.stack.name}/#{self.name}"},
+        {:affinity => "service!=#{self.stack.name}/#{self.name}"},
+        {:affinity => {:$regex => /^container(==|!=)#{self.name_with_stack}-(%s|\d+)$/}},
```

Comment:
The match for `container` affinities used to be missing completely, this now uses `name_with_stack` to get the `stack.service-X` naming.

## Testing

```yaml
stack: test/container-affinity
services:
  redis:
    image: redis
  bar:
    image: redis
    affinity:
      - container==$STACK.redis-1
```

```
I, [2017-10-31T15:11:48.622084 #12]  INFO -- StackDeployWorker: deploying service development/container-affinity/redis...
I, [2017-10-31T15:11:48.688947 #12]  INFO -- GridServiceSchedulerWorker: starting development/container-affinity/redis deploy
...
I, [2017-10-31T15:11:54.036320 #12]  INFO -- GridServiceDeployer: service development/container-affinity/redis has been deployed
I, [2017-10-31T15:11:54.053083 #12]  INFO -- GridServiceSchedulerWorker: deploying dependent service development/container-affinity/bar of deployed service development/container-affinity/redis
I, [2017-10-31T15:11:54.088635 #16]  INFO -- GridServiceSchedulerWorker: starting development/container-affinity/bar deploy
....
I, [2017-10-31T15:11:54.307250 #12]  INFO -- StackDeployWorker: skipping deployment of development/container-affinity/bar because it will be deployed by dependencies
...
I, [2017-10-31T15:11:59.849990 #16]  INFO -- GridServiceDeployer: service development/container-affinity/bar has been deployed
```